cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh

```diff
@@ -0,0 +1,862 @@
+        T& val = val_arr_[i];
+        T& buf = val_buf_[max_buf_len_ - max_arr_len_ + i];
+        if (is_greater_than<greater>(buf, val)) {
+          val         = buf;
+          idx_arr_[i] = idx_buf_[max_buf_len_ - max_arr_len_ + i];
+        }
+      }
+    } else {
+      for (int i = 0; i < max_arr_len_; ++i) {
+        if (is_greater_than<greater>(val_buf_[i], val_arr_[i])) {
+          val_arr_[i] = val_buf_[i];
+          idx_arr_[i] = idx_buf_[i];
+        }
+      }
+    }
+
+    bitonic_merge<capacity, !greater>::run(val_arr_, idx_arr_);
+
+    buf_len_ = 0;
+    set_k_th_();  // contains sync
+    for (int i = 0; i < max_buf_len_; ++i) {
+      val_buf_[i] = dummy_;
+    }
+  }
+
+  using WarpSort<capacity, greater, T, IdxT>::max_arr_len_;
+  using WarpSort<capacity, greater, T, IdxT>::val_arr_;
+  using WarpSort<capacity, greater, T, IdxT>::idx_arr_;
+  using WarpSort<capacity, greater, T, IdxT>::lane_;
+  using WarpSort<capacity, greater, T, IdxT>::k_;
+  using WarpSort<capacity, greater, T, IdxT>::dummy_;
+
+  static constexpr int max_buf_len_ = (capacity <= 64) ? 2 : 4;
+
+  T val_buf_[max_buf_len_];
+  IdxT idx_buf_[max_buf_len_];
+  int buf_len_;
+
+  T k_th_;
+  const int k_th_lane_;
+};
+
+template <int capacity, bool greater, typename T, typename IdxT>
+class WarpBitonic : public WarpSort<capacity, greater, T, IdxT> {
+ public:
+  __device__ WarpBitonic(int k, T dummy)
+    : WarpSort<capacity, greater, T, IdxT>(k, dummy), buf_len_(0)
+  {
+    for (int i = 0; i < max_arr_len_; ++i) {
+      val_buf_[i] = dummy_;
+    }
+  }
+
+  __device__ void add(const T* in, IdxT start, IdxT end)
+  {
+    add_first_(in, start, end);
+    start += capacity;
+    while (start < end) {
+      add_extra_(in, start, end);
+      merge_();
+      start += capacity;
+    }
+  }
+
+  __device__ void add(const T* in, const IdxT* in_idx, IdxT start, IdxT end)
+  {
+    add_first_(in, in_idx, start, end);
+    start += capacity;
+    while (start < end) {
+      add_extra_(in, in_idx, start, end);
+      merge_();
+      start += capacity;
+    }
+  }
+
+  __device__ void add(T val, IdxT idx)
+  {
+    for (int i = 0; i < max_arr_len_; ++i) {
+      if (i == buf_len_) {
+        val_buf_[i] = val;
+        idx_buf_[i] = idx;
+      }
+    }
+
+    ++buf_len_;
+    if (buf_len_ == max_arr_len_) {
+      bitonic_sort<capacity, greater>::run(val_buf_, idx_buf_);
+      merge_();
+
+      for (int i = 0; i < max_arr_len_; ++i) {
+        val_buf_[i] = dummy_;
+      }
+      buf_len_ = 0;
+    }
+  }
+
+  __device__ void done()
+  {
+    if (buf_len_ != 0) {
+      bitonic_sort<capacity, greater>::run(val_buf_, idx_buf_);
+      merge_();
+    }
+  }
+
+ private:
+  __device__ void add_first_(const T* in, IdxT start, IdxT end)
+  {
+    IdxT idx = start + lane_;
+    for (int i = 0; i < max_arr_len_; ++i, idx += WarpSize) {
+      if (idx < end) {
+        val_arr_[i] = in[idx];
+        idx_arr_[i] = idx;
```

Comment:
If we did not have indices in the beginning, then do we really need to generate one?

Reply:
Yes, for kNN we need to find out distances between which elements are the smallest in the end. Sometimes we don't need to pass indices in the first place, because `0..n-1` is implied (dense inputs).